### ISSUE TITLE
SSL Connection

### ISSUE THREAD

**Issue by vsoch** [HazyResearch/deepdive#375]
Is there support for an SSL connection?

**Comment by vsoch** [comment]
Just putting some notes here - I'm not sure if this is a bug or that I didn't (yet) install the system dependencies. [Here](https://github.com/HazyResearch/deepdive/blob/master/shell/load-db-driver.sh#L22) the different driver folders are expected to be under "util," but they are actually under "shell"

**Comment by chrismre** [comment]
@netj @SenWu @zhangce I don't see why this would be an issue...

**Comment by vsoch** [comment]
Great, thanks so much! If we figure it out, I'd be happy to write up a little documentation for it.

**Comment by chrismre** [comment]
@SenWu Can you look at this?

**Comment by SenWu** [comment]
Sure.

**Comment by SenWu** [comment]
@vsoch Let me rephrase your question. Do you mean whether we can configure Git to accept the our particular signed server certificate for deepdive repo (https remote)?

**Comment by vsoch** [comment]
@SenWu there isn't any issue with git - the issue is when using deepdive and connecting to the database (eg, any command with deepdive X" whether it be via a utils function to do initial database work, or a function embedded in the scala to work with the same database.) In the case of a url for postgres, it would be just added ?ssl=true. I'm hoping it will be as simple as adding an extra parser to read an ssl specification in the db.url, (and then parsed by the postres driver) and some module within the scala to do the same. I haven't worked with scala before, so I'm not much help, but here is a suggestion from one of the tech's at TACC:

      >> Actually, before going through all the trouble of rewriting code in a different language, I'd be interested in knowing if the following jdbc parameter would work in the case where we don't actually want to setup certificates.

      sslfactory=org.postgresql.ssl.NonValidatingFactory

And some more detail on the JDBC url:

      >> Coming in late to the conversation, so if this has already been discussed please forgive the extra mail, but with the JDBC you should be able to simply add properties to the JDBC URL. 

      String url = "jdbc:postgresql://localhost/test?user=fred&password=secret&ssl=true";
      Connection conn = DriverManager.getConnection(url);

https://jdbc.postgresql.org/documentation/80/connect.html

I played around with editing the driver functions and setting environment variables to expect SSL, but couldn't get anything to work.

**Comment by SenWu** [comment]
@vsoch Thanks for your clarification. We will add ssl option in next update.

**Comment by vsoch** [comment]
Fantastic, many thanks @SenWu.  There isn't huge rush because I was able to go through the entire walkthrough on my local machine, and am not held back in any way to start developing our methods locally. It's really a beautiful piece of software - I of course expect to have bugs and what not, but the workflow with the database, making extraction rules, and running pipelines is very intuitive! :L)

**Comment by chrismre** [comment]
You are far too nice!

We should be able to fix this soon... some of our team is very seriously
studying ... in hawaii for the week... I haven't seen them at the
conference, so I assume they are in their rooms coding :)

Chris

**Comment by vsoch** [comment]
oh yes, definitely :o)

![coding](https://imgs.xkcd.com/comics/compiling.png)

**Comment by SenWu** [comment]
@vsoch Can you try our new update in 'ssl_support' branch? In order to use ssl connections.You can specify ?ssl=true in the connection string, e.g., 
  jdbc:postgresql://localhost:5432/dbname?ssl=true
And Deepdive will connection to database via ssl connection with the non-validating ssl factory.

**Comment by vsoch** [comment]
Thanks @SenWu! I will give it a try asap - there might be some delay because my Brother is getting married tomorrow! :wedding:

**Comment by vsoch** [comment]
@SunWu, we are making progress! The ssl part seems to be working, and there are some additional bugs that are likely related to my installation (that I will look into). Here is how I tested, despite some installation bugs: 

First, I tried it as a user would, from the project directory:

      deepdive init-db

And here was the error:

      db-parse operation not available for "postgresql
      db-init operation not available for "postgresql
      db-execute operation not available for "postgresql
      db-query operation not available for "postgresql
      ~/local/util/load-db-driver.sh: line 61: db-parse: command not found
      ~/local/util/load-db-driver.sh: eval: line 37: unexpected EOF while looking for matching `"'

And this happens because there must be some issue adding the driver to the path in ~/local/util/load-db-driver.sh. I didn't look into this, instead I decided to keep going and try parsing the url:

      cd ~/local/util/
      url="postgresql://vsochat@xxx.xxxxxxxx.xxx.xxxx.edu:5432/database_name?ssl=true"
      ~/local/util/driver.postgresql/db-parse $url
      
Here it spit out all the variables into the window, and I copy pasted to execute, and the only issue being that the DEEPDIVE_JDBC_URL needed quotes since it had a "?" in it.

I'm not allowed to create the database itself (so I can't test dbinit) but I can execute stuff, so I decided to test that:

      cd driver.postgresql
      db-execute "\d+"
      Password for user vsochat: 
      No relations found.

Boom! Success! So - here are my thoughts. The ssl issue is definitely resolved. Setting PGSSLMODE to require and specifying ?ssl=true for the DEEPDIVE_JDBC_URL were spot on! I don't think that you need to do any more debugging - I want to do the entire thing from start to end fresh. I'm not convinced that some of my local changes aren't still lurking in code somewhere and mucking something up, because it works perfectly on my local machine. I am finishing up work for today, but I will be able to test this out this weekend. I will post an update!

**Comment by chrismre** [comment]
Adding @SenWu. We should doc this.

**Comment by SenWu** [comment]
@vsoch Happy to know it works.
@chrismre Doing doc now.

**Comment by SenWu** [comment]
Added doc in https://github.com/HazyResearch/deepdive/commit/4d39e685af090af776fbca5ccb0fa15a9cbcb602.

**Comment by vsoch** [comment]
ok, I can confirm with a fresh install that the previously mentioned bugs to use the driver are resolved! Given that I don't have permissions to create the database, the standard initdb commands don't work, however I can write custom scripts to do the database setup and loading. Thanks again for your help with this, and I'll be sure to post if I run into any trouble!
